Sovereign Chains Cross-Chain Execution (#1033)
multiversx/mx-docs: docs/sovereign/from-sovereign.md

Reviewer: I would say you have to explain header-verifier here and the endpoints used in this contract, in this page

# Execution starting from a Sovereign Chain
![From Sovereign](../../static/sovereign/from-sovereign.png)

For the user — whether an External Owned Account (EOA) like a user wallet or another smart contract — procedure is simple. An address will be able to perform a multiTransfer deposit for various types of tokens:
1. Fungible Tokens
2. Non-Fungible Tokens
3. Semi-Fungible Tokens
4. Meta ESDT Tokens

When making the deposit, the user specifies:
1. A destination address on the Sovereign Chain
2. The function to call
3. Any required arguments
4. *gasLimit* if a smart contract execution is needed after the cross-chain operation

> Each action that can be executed remotely through this contract is called an *Operation*. The endpoint responsible for executing those operations is called `execute_operations`.

Reviewer: Each deposit in ESDT-safe in Sovereign will be an outgoing operation which will be sent to the bridge service which will call header-verifier and ESDT-safe in mainchain. Rephrase something like that.

Reviewer: Also, you can link the `deposit` endpoint from the other page somewhere here

> #### Execution from inside the Sovereign Chain flow

Reviewer: This is `Sovereign Chain -> Main Chain transfer flow`

> 1. User sends token to the ESDT-Safe smart contract on Sovereign.

Reviewer: User deposits tokens in the ESDT-Safe...

2. The validators generate a proof on the Sovereign Chain for a batch of transfers, which will be sent to the mainchain ESDT-Safe contract.
3. Validators add this information to the sovereignChainBlockBody. Otherwise the block is not signed.

> 4. Leader will push the created txData to the mainchain

Reviewer: Outgoing operations will be created at the end of the round and will be signed by all the validators and sent to the bridge service by the leader.

> 5. The ESDT-Safe contract on the mainchain verifies the proof and executes the transfers.

Reviewer: Bridge service will send the operations to to header-verifier for verification then to ESDT-Safe for execution.

> 6. Sovereign validators notarize the completion of the transfer in the subsequent Sovereign block by receiving the attestation log event directly from the mainchain.

Reviewer: At the end of the execution success/fail, a confirmation event will be added which will be received in sovereign through the observer and then the cross chain transfer will be completed

Before talking about the logic of the endpoint, let’s give some more context about the *Operation*.

This structure, derived with multiple traits for serialization, encoding, and type interfacing, represents a fundamental unit of execution within the bridging mechanism.

:::note
The source code for the following structures can be found [here](https://github.com/multiversx/mx-sovereign-sc/blob/main/common/transaction/src/lib.rs)
:::

```rust
#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, ManagedVecItem, Clone)]
pub struct Operation<M: ManagedTypeApi> {
    pub to: ManagedAddress<M>,
    pub tokens: ManagedVec<M, OperationEsdtPayment<M>>,
    pub data: OperationData<M>,
}
```

- `to`: specifies the destination of the *Operation*
- `tokens`: represents one or more token transfers associated with the operation
- `data`: encapsulates additional instructions or parameters that guide the execution of the operation

> Why do we need so many custom structures? This a question related to clean code principles, the answer is easy: to ensure that complex logic remains organized and maintainable, thus improving code extension and flexibility.

Reviewer: I would delete this phrase

```rust
pub struct OperationEsdtPayment<M: ManagedTypeApi> {
    pub token_identifier: TokenIdentifier<M>,
    pub token_nonce: u64,
    pub token_data: EsdtTokenData<M>,
}
```

This struct describes a single token transfer action within an Operation. Each Operation can have one or more of such payments, with that enabling the transfer of a variety of tokens during a cross-chain transaction.

- `token_identifier`: used for the identification of the token
- `token_nonce`: if the token is Non-Fungible or Semi-Fungible, it will have a custom nonce, if not the value will be 0

> - `token_data`: a structure holding metadata and additional token-related details used for minting, token creation (for more details about this structure, please visit the official ESDT Documentation)

Reviewer: You can use this reference for token data `/developers/developer-reference/sc-api-functions#get_esdt_token_data`

```rust
pub struct OperationData<M: ManagedTypeApi> {
    pub op_nonce: TxId,
    pub op_sender: ManagedAddress<M>,
    pub opt_transfer_data: Option<TransferData<M>>,
}
```

`OperationData` encapsulates the needed information for the *Operation* that needs to be executed. This isn’t just another data definition, we’ve already seen data-related fields elsewhere. Instead, it centralizes the contextual information that *Operation* needs before, during, and after execution.

- `op_nonce`: is used for the identification of each *Operation*
- `op_sender`: represents the original sender of the *Operation*
- `opt_transfer_data`: an optional `TransferData` field, when present, contains details about the cross-chain execution of another Smart Contract

```rust
pub struct TransferData<M: ManagedTypeApi> {
    pub gas_limit: GasLimit,
    pub function: ManagedBuffer<M>,
    pub args: ManagedVec<M, ManagedBuffer<M>>,
}
```

> `TransferData` represents the description of the remove execution of another Smart Contract, one standout feature of the `executeBridgeOps` endpoint is the capability of executing multiple endpoint calls in one transaction.

Reviewer: - "remove execution"?
- "executing multiple endpoint calls in one transaction" this is not true

- `gas_limit`: specifies the needed gas for the execution of all other endpoints.
- `function`: the name of the endpoint that will be executed.
- `args`: the arguments for the calls.

:::note
The source code for the endpoint can be found [here](https://github.com/multiversx/mx-sovereign-sc/blob/main/esdt-safe/src/from_sovereign/transfer_tokens.rs)
:::
### Executing an *Operation*

> This endpoint serves as the key component in orchestrating cross-chain operations within the bridging mechanism.

Reviewer: I don't really understand this sentence. This endpoint is used for executing the operation....

```rust
#[endpoint(executeBridgeOps)]
 fn execute_operations(
    &self, 
    hash_of_hashes: ManagedBuffer, 
    operation: Operation<Self::Api>
)
```

> - `hash_of_hashes`: the key for the hashes map used to track executed and non executed Operations

Reviewer: This is the hash of all the hashes of the operations that were sent in a round

> - `operation`: the data and the behavior for the cross-chain execution that must happen

Reviewer: I'm not sure about the description for `operation`

> 1. If the current chain isn’t a Sovereign Chain, execution must happen inside the Main Chain.

Reviewer: I would say you can delete this point

> 2. Verify that the given Operation’s hash is registered by the Header-Verifier Smart Contract.

Reviewer: Just explain what happens in the `execute_operations` endpoint
- calculate operation hash
- verify if registered and lock operation hash
- mint tokens or get from account
- distribute tokens
- emit confirmation event and fail event if needed

# Execution starting from a Sovereign Chain
![From Sovereign](../../static/sovereign/from-sovereign.png)

> 

Reviewer 2: First specify all the capabilities:
The solution enables users will be able to *transfer tokens* and *execute SC* from sovereign to main chain.
Then you start to describe each case (transfer, execute) below

> For the user — whether an External Owned Account (EOA) like a user wallet or another smart contract — procedure is simple. An address will be able to perform a multi tokens transfer with various types of tokens:

Reviewer 2: Remove "For the user"

# Execution starting from a Sovereign Chain
![From Sovereign](../../static/sovereign/from-sovereign.png)

> For the user — whether an External Owned Account (EOA) like a user wallet or another smart contract — procedure is simple. An address will be able to perform a multiTransfer deposit for various types of tokens:

Reviewer: ```suggestion
For the user — whether an External Owned Account (EOA) like a user wallet or another smart contract — procedure is simple. An address will be able to perform a multi tokens transfer with various types of tokens:
```

# Execution starting from a Sovereign Chain
![From Sovereign](../../static/sovereign/from-sovereign.png)

For the user — whether an External Owned Account (EOA) like a user wallet or another smart contract — procedure is simple. An address will be able to perform a multiTransfer deposit for various types of tokens:
- Fungible Tokens
- (Dynamic) Non-Fungible Tokens
- (Dynamic) Semi-Fungible Tokens
- (Dynamic) Meta ESDT Tokens

When making the deposit, the user specifies:
1. A destination address on the Sovereign Chain
2. `TransferData` if the execution contains a smart contract call, which contains gas, function and arguments

For each deposit in the ESDT-Safe smart contract inside a Sovereign Chain, an outgoing operation will be sent to the bridge service that calls the Header-Verifier and ESDT-Safe contracts in the MultiversX MainChain.

:::note
Here is the [link](https://github.com/multiversx/mx-sovereign-sc/blob/main/esdt-safe/src/to_sovereign/create_tx.rs) to the `deposit` endpoint
:::

Each action that can be executed remotely through this contract is called an *Operation*. The endpoint responsible for executing those operations is called `execute_operations`.

> #### Execution from inside the Sovereign Chain to the MainChain flow

Reviewer: To keep it consistent you can say:
```suggestion
## Sovereign Chain to Main Chain transfer flow
```

1. User deposits token to the ESDT-Safe smart contract on Sovereign.
2. Outgoing *Operations* are created at the end of the round.
3. Validators sign all the outgoing *Operations*.
4. Leader sends *Operations* to the bridge service.
5. Bridge service sends the *Operations* to the Header-Verifier for registration and verification, and then to ESDT-Safe for execution.
6. At the end of the execution success/fail, a confirmation event will be added which will be received in sovereign through the observer and then the cross chain transfer will be completed.

> ### Header-Verifier SC

Reviewer: ```suggestion
## Header-Verifier SC
```

As mentioned in the [Introduction](cross-chain-execution.md) the Header-Verifier smart contract is to verify signatures, store the *BLS Keys* of the validators and register incoming *Operations*.

> 

Reviewer: ```suggestion
### Registering a set of *Operations*

```

```rust
    #[endpoint(registerBridgeOps)]
    fn register_bridge_operations(
        &self,
        signature: BlsSignature<Self::Api>,
        bridge_operations_hash: ManagedBuffer,
        operations_hashes: MultiValueEncoded<ManagedBuffer>,
    )
```

> Any transaction before being executed has to be registered in this smart contract. The reason behind registering any incoming *Operation* is to create a history of all already registered *Operations* and with that an already registered *Operation* doesn't need to be re-registered and re-executed by the validators.

Reviewer: ```suggestion
Any *Operation* before being executed has to be registered in this smart contract. The reason behind this is that the hash will be verified and it will be locked until the operation is executed by the ESDT-Safe contract.
```

> The registering endpoint follows this flow:
> 1. If the incoming `bridge_operations_hash` is not in the `hash_of_hashes_history` storage mapper, otherwise, the endpoint will return a panic.
> 2. If the hash of all `operations_hashes` and `bridge_operations_hash` are equal, otherwise, the endpoint will return a panic.
> 3. All the `operations_hashes` are registered in the smart contract's storage with the `OperationsHashStatus::NotLocked` status.
> 4. The `bridge_operations_hash` is inserted in the `hash_of_hashes_history` storage mapper.

Reviewer: ```suggestion
The registering endpoint operates as follows:
1. Verifies that  `bridge_operations_hash` is not found in the `hash_of_hashes_history` storage mapper, otherwise it will return an error.
2. Verifies that the hash of all `operations_hashes` matches the `bridge_operations_hash, otherwise, the endpoint will return an error.
3. All `operations_hashes` are stored in the smart contract's storage with the status OperationsHashStatus::NotLocked.
4. The `bridge_operations_hash` is added to the `hash_of_hashes_history` storage mapper.
```

```

The Header-Verifier has a system in place for locking *Operation* hashes. Locking those registered hashes prevents any unwanted behaviour when executing or removing an *Operation* hash. Remember that the execution of *Operations* can only be done by the ESDT-Safe smart contract. This endpoint when called will follow this flow:

1. Check if the caller is the ESDT-Safe smart contract.
2. Check if the *Operation* is registered.
3. If the hash is not locked set the status in the storage as locked or else return panic.

:::note
The hash can be in two different states: `OperationHashStatus::NotLocked` or `OperationHashStatus::Locked`
:::

```rust
    #[endpoint(removeExecutedHash)]
    fn remove_executed_hash(&self, hash_of_hashes: &ManagedBuffer, operation_hash: &ManagedBuffer)
```

After registering and executing an *Operation* the status of the hash associated to it must be removed from the Header-Verifier's internal storage. This endpoint will be called by the ESDT-Safe smart contract after the execution of the *Operation* is successful. The steps are pretty clear:

1. Check if the caller is the ESDT-Safe smart contract.
2. Remove the status of the hash from storage.

> 
> Before talking about the logic of the endpoint, let’s give some more context about the *Operation*.

Reviewer: ```suggestion
## ESDT-Safe SC
```

:::

```rust
#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, ManagedVecItem, Clone)]
pub struct Operation<M: ManagedTypeApi> {
    pub to: ManagedAddress<M>,
    pub tokens: ManagedVec<M, OperationEsdtPayment<M>>,
    pub data: OperationData<M>,
}
```

- `to`: specifies the destination of the *Operation*
- `tokens`: represents one or more token transfers associated with the operation
- `data`: encapsulates additional instructions or parameters that guide the execution of the operation

```rust
pub struct OperationEsdtPayment<M: ManagedTypeApi> {
    pub token_identifier: TokenIdentifier<M>,
    pub token_nonce: u64,
    pub token_data: EsdtTokenData<M>,
}
```

This struct describes a single token transfer action within an *Operation*. Each Operation can have one or more of such payments, with that enabling the transfer of a variety of tokens during a cross-chain transaction. 

- `token_identifier`: used for the identification of the token
- `token_nonce`: if the token is Non-Fungible or Semi-Fungible, it will have a custom nonce, if not the value will be 0

> - `token_data`: a structure holding metadata and additional token-related details used for minting, token creation (for more details about this structure, please visit the official ESDT Documentation)
> 
> :::note
> You can use the [System SC API](../developers/developer-reference/sc-api-functions#get_esdt_token_data) method as reference.
> :::

Reviewer: Let's keep it simple

```suggestion
- `token_data`: a structure holding metadata and other token properties
```
